fcrepo4/fcrepo4
# Adds support for acl:Control predicate.

Resolves:

**Adds support for acl:Control predicate.**
* * *

**JIRA Ticket**:  https://jira.duraspace.org/browse/FCREPO-2743

* Other Relevant Links (Mailing list discussion, related pull requests, etc.)

# What does this Pull Request do?
Adds support for acl:Control predicate.  It includes extensive unit tests but no automated integration tests.

# What's new?
The acl:Control predicate is now respected.  In order to GET, HEAD, PUT, PATCH or DELETE an acl, the effective acl for the acl's resource must contain an acl:Control predicate.

# How should this be tested?
1.  create a control acl.
```
@prefix acl: <http://www.w3.org/ns/auth/acl#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

<#restricted> a acl:Authorization ;
acl:agent "testuser" ;
acl:mode acl:Control ;
acl:accessTo </rest/control> .
```
2.  create a readwrite acl.
```
@prefix acl: <http://www.w3.org/ns/auth/acl#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

<#restricted> a acl:Authorization ;
acl:agent "testuser" ;
acl:mode acl:Read;
acl:mode acl:Write;
acl:accessTo </rest/readwrite> .
```
3. create a resource and give it an acl with acl:Control
 
```
curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/control
curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/control/fcr:acl     --data-binary "@control.ttl" -H "Content-Type: text/turtle"
```
4. create a resource and give it an acl with acl:Read and Write

curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/readwrite
curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/readwrite/fcr:acl     --data-binary "@readwrite.ttl" -H "Content-Type: text/turtle"
```
5. create  a child of the readwrite resource:
```
curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/readwrite/child
```
6. try to access the child's acl as testuser and verify 403 response
```
curl -u testuser:testpass -v -X GET http://localhost:8080/rest/readwrite/child/fcr:acl
curl -u testuser:testpass -v -X HEAD http://localhost:8080/rest/readwrite/child/fcr:acl
curl -u testuser:testpass -v -X PUT http://localhost:8080/rest/readwrite/child/fcr:acl     --data-binary "@readwrite.ttl" -H "Content-Type: text/turtle"
curl -u testuser:testpass -v -X DELETE http://localhost:8080/rest/readwrite/child/fcr:acl     --data-binary "@readwrite.ttl" -H "Content-Type: text/turtle"
```

7. create  a child of the control resource:
```
curl -u fedoraAdmin:fedoraAdmin -v -X PUT http://localhost:8080/rest/control/child
```
8. try to access the child's acl as testuser and verify 404 response
```
curl -u testuser:testpass -v -X GET http://localhost:8080/rest/control/child/fcr:acl
curl -u testuser:testpass -v -X HEAD http://localhost:8080/rest/control/child/fcr:acl

```
9 create child-control.ttl
```
@prefix acl: <http://www.w3.org/ns/auth/acl#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

<#restricted> a acl:Authorization ;
acl:agent "testuser" ;
acl:mode acl:Control ;
acl:accessTo </rest/control/child> .

```
10. try to create the child's acl as testuser and verify 201 response

```
curl -u testuser:testpass -v -X PUT http://localhost:8080/rest/control/child/fcr:acl     --data-binary "@child-control.ttl" -H "Content-Type: text/turtle"
```
11. try to patch the child's acl as testuser and verify 204

```
curl -u testuser:testpass -v -X PATCH http://localhost:8080/rest/control/child/fcr:acl     --data  "PREFIX acl: <http://www.w3.org/ns/auth/acl#>  INSERT { <#restricted> acl:mode acl:Read } WHERE { }" -H "Content-Type: application/sparql-update"
```
12. verify that change was made.
```
curl -u testuser:testpass -v -X GET http://localhost:8080/rest/control/child/fcr:acl
```

13. try to delete the child's acl as testuser and verify 204 response
```
curl -u testuser:testpass -v -X DELETE http://localhost:8080/rest/control/child/fcr:acl
```

# Additional Notes:
* There are no  integration tests for this PR.  
* Does this change require documentation to be updated? Yes 
* Does this change add any new dependencies? No
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? No
* Could this change impact execution of existing code? Yes

# Interested parties
@awoods, @peichman-umd @fcrepo4/committers
